### PR TITLE
Affiliate support

### PR DESCRIPTION
**Pull request by dicksonpys**

### CODE REVIEW

File: ts-client/src/vault/index.ts [hunk 1]
```diff
@@ -74,13 +88,29 @@ export default class VaultImpl implements VaultImplementation {
...
+            AffiliateIDL as AffiliateVaultIdl,
+            opt?.affiliateProgramId || AFFILIATE_PROGRAM_ID,
+            provider
+          )
+          : undefined
+      }
+    );
   }
 
   public async getUserBalance(owner: PublicKey): Promise<BN> {
```
Comment (andrewsource147) [inline ts-client/src/vault/index.ts:116]: we need to update user balance function if they are depositing though affiliate program
https://github.com/mercurial-finance/vault-periphery/blob/main/affiliate/rust-client/src/user.rs#L136
`user_state.lp_token`

File: ts-client/src/vault/index.ts [hunk 2]
```diff
@@ -157,25 +184,146 @@ export default class VaultImpl implements VaultImplementation {
...
     }
+
+    return {
+      preInstructions,
+      userToken,
+      userLpToken,
+    };
+  }
+
+  private async createAffiliateATAPreInstructions(owner: PublicKey) {
```
Comment (andrewsource147) [inline ts-client/src/vault/index.ts:195]: We need a function to get partner info
https://github.com/mercurial-finance/vault-periphery/blob/main/affiliate/rust-client/src/partner.rs#L45

Comment (dicksonpys) [reply ts-client/src/vault/index.ts:206]: What does this partner info looks like? Is it this one?
```ts
 {
      "parsed": {
        "info": {
          "isNative": true,
          "mint": "So11111111111111111111111111111111111111112",
          "owner": "7236FoaWTXJyzbfFPZcrzg3tBpPhGiTgXsGWvjwrYfiF",
          "rentExemptReserve": {
            "amount": "2039280",
            "decimals": 9,
            "uiAmount": 0.00203928,
            "uiAmountString": "0.00203928"
          },
          "state": "initialized",
          "tokenAmount": {
            "amount": "0",
            "decimals": 9,
            "uiAmount": 0,
            "uiAmountString": "0"
          }
        },
        "type": "account"
      },
      "program": "spl-token",
      "space": 165
    }
```

I can add this to the demo

File: ts-client/src/vault/index.ts [hunk 3]
```diff
@@ -157,25 +184,146 @@ export default class VaultImpl implements VaultImplementation {
...
+
+    let preInstructions: TransactionInstruction[] = [];
+
+    let partnerAddress: PublicKey | undefined;
+    let userAddress: PublicKey | undefined;
+    let userToken: PublicKey | undefined;
+    let userLpToken: PublicKey | undefined;
+
+    // Withdraw with Affiliate
+    if (this.affiliateId && this.affiliateProgram) {
```
[the remark this answers is not in the document]
Comment (00xSam) [reply ts-client/src/vault/index.ts:268]: ahh, typescript 🤦🏿‍♂️

File: ts-client/src/vault/tests/affiliate.test.ts [hunk 4]
```diff
@@ -59,6 +63,10 @@ describe('Interact with Vault in devnet', () => {
         const depositResult2 = await provider.sendAndConfirm(depositTx2);
         console.log('Deposit result', depositResult2);
         expect(typeof depositResult2).toBe('string');
+
+        // Check balance again
+        const userBalanceDeposit2 = await vaultImpl.getUserBalance(mockWallet.publicKey);
+        expect(Number(userBalanceDeposit2)).toBeGreaterThan(0);
```
Comment (00xSam) [inline ts-client/src/vault/tests/affiliate.test.ts:69]: i think here we can check greater than the first amount?

File: ts-client/src/vault/index.ts (hunk 2 above)
Comment (andrewsource147) [reply ts-client/src/vault/index.ts:206]: No it should be like
```
Partner { partner_token: 6EqwccrRKaVqhDijnNvyz9Hh919fsJ2G7JDB9M9HjQuZ, vault: FERjPVNEa7Udq8CEv68h6tPL46Tq7ieE49HrE2wea3XT, total_fee: 2138, fee_ratio: 2000, cummulative_fee: 0 }
```

Comment (andrewsource147) [reply ts-client/src/vault/index.ts:206]: You can refer partner struct in IDL

File: ts-client/src/vault/tests/affiliate.test.ts (hunk 4 above)
Comment (dicksonpys) [reply ts-client/src/vault/tests/affiliate.test.ts]: yep done both, good suggestion

**Review by 00xSam — APPROVED** [review]
LGTM

**Pull request merged**